File: 7-permutations.py
"""
permutation
nPr = n! / (n - r)!

nums = [1, 2, 3]

1
  2
    3 => [1, 2, 3]
  3
    2 => [1, 3, 2]
2
  1
    3 => [2, 1, 3]
  3
    1 => [2, 3, 1]
3
  1
    2 => [3, 1, 2]
  2
    1 => [3, 2, 1]
"""


def permute(nums):
    """
    0(n!) time
    0(n^2) space
    """
    permutations = []

    def dfs(picked, unpicked):
        if not unpicked:
            return permutations.append(picked)
        for i, num in enumerate(unpicked):
            dfs(picked + [num], unpicked[:i] + unpicked[i + 1 :])

    dfs([], nums)

    print(permutations)
    return permutations


permute([1, 5, 6])

File: 7-permutations.js
function permute(nums) {
	const result = [];

	function dfs(picked, unpicked) {
		if (unpicked.length === 0) {
			result.push(picked);
			return;
		}

		for (let i = 0; i < unpicked.length; i++) {
			dfs([...picked, unpicked[i]], [...unpicked.slice(0, i), ...unpicked.slice(i + 1)]);
		}
	}

	dfs([], nums);
	console.log(result);
	return result;
}

permute([3, 5, 8]);
